File: backend/system_design/api_types.py
from ninja import Schema
from typing import List, Optional
from datetime import datetime, timedelta
from uuid import UUID


class SDCourseResponse(Schema):
    id: UUID
    name: str
    description: str
    order: int
    is_active: bool
    lessons_count: Optional[int] = 0


class SDLessonResponse(Schema):
    id: UUID
    name: str
    description: str
    order: int
    learning_objectives: List[str] = []
    reference_diagram: Optional[dict] = None


class SDCourseDetailResponse(Schema):
    id: UUID
    name: str
    description: str
    order: int
    is_active: bool
    lessons: List[SDLessonResponse] = []


class SDLessonProgressResponse(Schema):
    id: UUID
    lesson: SDLessonResponse
    conversation_id: Optional[UUID] = None
    status: str
    ai_confidence: float = 0.0
    covered_points: List[str] = []
    remaining_points: List[str] = []
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None


class UserSDCourseResponse(Schema):
    id: UUID
    course: SDCourseResponse
    conversation_id: Optional[UUID] = None
    current_lesson: Optional[SDLessonResponse] = None
    progress_percentage: float = 0.0
    is_completed: bool = False
    started_at: datetime
    completed_at: Optional[datetime] = None
    is_active: bool = True
    progress: List[SDLessonProgressResponse] = []


class SDCaseStudySummary(Schema):
    id: UUID
    title: str
    slug: str
    difficulty: str
    topics: List[str] = []
    overview: str = ""


class SDCaseStudyDetail(SDCaseStudySummary):
    functional_requirements: List[str] = []
    non_functional_requirements: List[str] = []
    capacity: dict = {}
    key_components: List[dict] = []
    tradeoffs: List[str] = []
    reference_diagram: Optional[dict] = None


class CreateSDPracticeSchema(Schema):
    case_study_id: UUID


class SDPracticeSessionResponse(Schema):
    id: UUID
    case_study: SDCaseStudyDetail
    conversation_id: UUID
    current_phase: int = 1
    phase_states: dict = {}
    weak_areas: List[str] = []
    status: str = "in_progress"
    started_at: datetime
    completed_at: Optional[datetime] = None


class ComponentTutorCreate(Schema):
    kind: str


class ComponentTutorResponse(Schema):
    conversation_id: UUID


class MarkComponentSchema(Schema):
    kind: str


class ComponentProgressResponse(Schema):
    component_kind: str
    completed_at: datetime

File: frontend/src/types/system_design/api_types.ts
import type { ReactFlowDiagram } from "../ai_core/api_types";

export type SDProgressStatus = "not_started" | "learning" | "completed";

export interface SDCourseResponse {
  id: string;
  name: string;
  description: string;
  order: number;
  is_active: boolean;
  lessons_count: number;
}

export interface SDLessonResponse {
  id: string;
  name: string;
  description: string;
  order: number;
  learning_objectives: string[];
  reference_diagram: ReactFlowDiagram | null;
}

export interface SDCourseDetailResponse {
  id: string;
  name: string;
  description: string;
  order: number;
  is_active: boolean;
  lessons: SDLessonResponse[];
}

export interface SDLessonProgressResponse {
  id: string;
  lesson: SDLessonResponse;
  conversation_id: string | null;
  status: SDProgressStatus;
  ai_confidence: number;
  covered_points: string[];
  remaining_points: string[];
  started_at: string | null;
  completed_at: string | null;
}

export interface UserSDCourseResponse {
  id: string;
  course: SDCourseResponse;
  conversation_id: string | null;
  current_lesson: SDLessonResponse | null;
  progress_percentage: number;
  is_completed: boolean;
  started_at: string;
  completed_at: string | null;
  is_active: boolean;
  progress: SDLessonProgressResponse[];
}

export interface SDLearningMessage {
  id: string;
  sender: "user" | "ai";
  content: string;
  timestamp: string;
  code_snippet: string | null;
  language: string | null;
  type?: string | null;
  next_action?: string | null;
  diagram?: ReactFlowDiagram | null;
}

export type SDCaseStudyDifficulty = "easy" | "medium" | "hard";

export interface SDCaseStudySummary {
  id: string;
  title: string;
  slug: string;
  difficulty: SDCaseStudyDifficulty;
  topics: string[];
  overview: string;
}

export interface SDCaseStudyDetail extends SDCaseStudySummary {
  functional_requirements: string[];
  non_functional_requirements: string[];
  capacity: Record<string, string>;
  key_components: { name: string; responsibility: string }[];
  tradeoffs: string[];
  reference_diagram: ReactFlowDiagram | null;
}

export type SDPracticeStatus = "in_progress" | "completed";

export interface SDPracticeSessionResponse {
  id: string;
  case_study: SDCaseStudyDetail;
  conversation_id: string;
  current_phase: number;
  phase_states: Record<string, unknown>;
  weak_areas: string[];
  status: SDPracticeStatus;
  started_at: string;
  completed_at: string | null;
}